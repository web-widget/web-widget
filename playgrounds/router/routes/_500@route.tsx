import './(css)/error.css';
import { defineRouteFallbackComponent } from '@web-widget/react';
import VueCounter from '@playgrounds/web-router-vue3/Counter@widget.vue?as=jsx';

export const fallback = defineRouteFallbackComponent(function Page500(ctx) {
  return (
    <main>
      <h1>⚠️ 500</h1>
      <pre>{ctx.message}</pre>
      <VueCounter count={3} />
    </main>
  );
});
