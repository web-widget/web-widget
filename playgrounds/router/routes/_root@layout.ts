import { html, render, unsafeHTML, unsafeStreamToHTML } from '@web-widget/html';

import type { HTML } from '@web-widget/html';
import type { LayoutComponentProps } from '@web-widget/web-router';
import { renderMetaToString } from '@web-widget/helpers';

export { render };

export default function RootLayout({
  meta,
  children,
}: LayoutComponentProps): HTML {
  /* eslint-disable prettier/prettier */
  return html`<!doctype html>
    <html lang="${meta.lang}">
      <head>
        ${unsafeHTML(renderMetaToString(meta))}
      </head>
      <body>
        ${children instanceof ReadableStream
          ? unsafeStreamToHTML(children)
          : unsafeHTML(children)}
      </body>
    </html>`;
}
