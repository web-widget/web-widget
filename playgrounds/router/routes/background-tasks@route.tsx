import { defineRouteComponent, defineRouteHandler } from '@web-widget/react';
import BaseLayout from './(components)/BaseLayout';

export const handler = defineRouteHandler({
  async GET(ctx) {
    ctx.waitUntil(
      new Promise((resolve) =>
        setTimeout(() => {
          console.log('Background task completed');
          resolve(undefined);
        }, 1000)
      )
    );
    return ctx.render();
  },
});

export default defineRouteComponent(function Page() {
  return (
    <BaseLayout>
      <h1>Background tasks</h1>
      <p>Please check the terminal log.</p>
    </BaseLayout>
  );
});
