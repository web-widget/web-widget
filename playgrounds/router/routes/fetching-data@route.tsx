import { defineRouteComponent, defineRouteHandler } from '@web-widget/react';
import { useLocation } from '@web-widget/helpers/navigation';
import type { HelloData } from './api/hello-world@route.ts';
import BaseLayout from './(components)/BaseLayout.tsx';
import ReactGithub from './(components)/Github@widget.tsx';
import VanillaGithub from './(components)/VanillaGithub@widget';
import VueGithub from '@playgrounds/web-router-vue3/Github@widget.vue?as=jsx';
import Vue2Github from '@playgrounds/web-router-vue2/Github@widget.vue?as=jsx';
import UserCard from './(components)/UserCard@widget.tsx';

async function fetchData(url: URL) {
  const data = await fetch(`${url.origin}/api/hello-world`);
  return (await data.json()) as HelloData;
}

export const handler = defineRouteHandler<HelloData>({
  async GET(ctx) {
    const data = await fetchData(useLocation());
    return ctx.html(data);
  },
});

export default defineRouteComponent<HelloData>(function Page({ data }) {
  return (
    <BaseLayout>
      <h1>Data Fetching Example</h1>

      <div style={{ display: 'grid', gap: '20px', marginTop: '20px' }}>
        <div>
          <h3>React Component</h3>
          <ReactGithub username="react" />
        </div>

        <div>
          <h3>Vue3 Component</h3>
          <VueGithub username="vuejs" />
        </div>

        <div>
          <h3>Vue2 Component</h3>
          <Vue2Github username="angular" />
        </div>

        <div>
          <h3>Vanilla JavaScript Component</h3>
          <VanillaGithub username="sveltejs" />
        </div>
      </div>

      <h2>Frontend Frameworks Showcase</h2>
      <p>
        Display mainstream frontend frameworks using beautiful card components:
      </p>

      <div
        style={{
          display: 'grid',
          gridTemplateColumns: 'repeat(auto-fit, minmax(300px, 1fr))',
          gap: '20px',
          marginTop: '20px',
        }}>
        <UserCard username="react" />
        <UserCard username="vuejs" />
        <UserCard username="angular" />
        <UserCard username="sveltejs" />
        <UserCard username="preactjs" />
        <UserCard username="solidjs" />
      </div>

      <h2>Data</h2>

      <div>
        <pre>{JSON.stringify(data, null, 2)}</pre>
      </div>
    </BaseLayout>
  );
});
