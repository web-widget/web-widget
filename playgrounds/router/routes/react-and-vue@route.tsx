import { defineRouteComponent, defineMeta } from '@web-widget/react';
import ReactCounter from './(components)/Counter@widget';
import VanillaCounter from './(components)/VanillaCounter@widget?as=jsx';
import BaseLayout from './(components)/BaseLayout.tsx';
import VueCounter from './(vue3)/Counter@widget.vue';
import Vue2Counter from '@playgrounds/web-router-vue2/Counter@widget.vue';
import { toReact } from './(vue3)/helpers';
import { toReact as vue2ToReact } from './(vue2)/helpers';

const RVueCounter = toReact(VueCounter);
const RVue2Counter = vue2ToReact(Vue2Counter);

export const meta = defineMeta({
  title: 'Hello, Web Widget',
});

export default defineRouteComponent(function Page() {
  return (
    <BaseLayout>
      <h1>Using react and vue together</h1>

      <h2>React component:</h2>
      <ReactCounter count={3} />

      <h2>Vue3 component:</h2>
      <RVueCounter count={3} />

      <h2>Vue2 component:</h2>
      <RVue2Counter count={3} />

      <h2>Vanilla component:</h2>
      <VanillaCounter count={3} />
    </BaseLayout>
  );
});
