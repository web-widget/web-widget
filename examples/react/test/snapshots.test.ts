import { describe, test, expect } from 'vitest';
import routes from '../routemap.server.json';
import fetch from './fetch';

const STATIC_ROUTES: { pathname: string }[] = routes.routes.filter(
  ({ pathname }) => !/[:(){}*+?]/.test(pathname)
);

STATIC_ROUTES.push({ pathname: '/examples/_404' });

// Replace local file paths with placeholders to ensure snapshots are consistent across different devices
function replaceLocalPaths(content: string): string {
  return (
    content
      // Replace all file:// protocol URLs
      .replace(/file:\/\/\/[^"'\s>]+/g, 'file:///LOCAL_PATH')
  );
}

describe('Should match snapshot', () => {
  test.each(STATIC_ROUTES.map((route) => [route.pathname]))(
    'Request "%s" should match snapshot',
    async (pathname) => {
      const result = await fetch(`${pathname}`);
      expect(result.status).toMatchSnapshot(`${pathname}@status`);
      expect(result.statusText).toMatchSnapshot(`${pathname}@statusText`);
      expect(Object.fromEntries(result.headers.entries())).toMatchSnapshot(
        `${pathname}@headers`
      );
      expect(replaceLocalPaths(await result.text())).toMatchSnapshot(
        `${pathname}@body`
      );
    }
  );
});
