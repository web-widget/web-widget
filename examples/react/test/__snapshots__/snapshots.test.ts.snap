// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot > Request "/" should match snapshot > /@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title >Home - Web Widget</title><style >web-widget{display:contents}</style><script type="module" src="/entry.client.ts"></script><script id="state:web-router" type="application/json">{"pathname":"/","params":{}}</script>
      </head>
      <body>
        <header class="_header_d366fc"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li><li><a href="/fetch">Fetching data</a></li><li><a href="/greet/web-widget">Dynamic routes</a></li><li><a href="/action">Server actions</a></li></ul></nav></header><main class="_main_d366fc"><h1>Welcome to Web Widget.</h1><p>Try to update this message in the ./routes/index@route.tsx file.</p><h2>React</h2><!--$--><web-widget import="/routes/(components)/Counter@widget.tsx" name="ReactCounter" data="{&quot;name&quot;:&quot;React Counter&quot;,&quot;start&quot;:0}" loading="lazy" rendertarget="light" recovering=""><div class="_counter_6a98f7"><button>−</button><span class="_count_6a98f7">0</span><button>+</button></div></web-widget><!--/$--><h2>Vue</h2><!--$--><web-widget import="/routes/(components)/Counter@widget.vue?as=jsx" name="VueCounter" data="{&quot;name&quot;:&quot;Vue3 Counter&quot;,&quot;start&quot;:0}" loading="lazy" rendertarget="light" recovering=""><div class="counter" data-v-a67ee2b2><button data-v-a67ee2b2>−</button><span class="count" data-v-a67ee2b2>0</span><button data-v-a67ee2b2>+</button></div></web-widget><!--/$--></main><footer class="_footer_d366fc"><p>This is a footer</p></footer>
      </body>
    </html>"
`;

exports[`Should match snapshot > Request "/" should match snapshot > /@headers 1`] = `
{
  "content-type": "text/html; charset=utf-8",
  "x-module-source": "source://routes/index@route.tsx",
  "x-powered-by": "@web-widget/web-router",
}
`;

exports[`Should match snapshot > Request "/" should match snapshot > /@status 1`] = `200`;

exports[`Should match snapshot > Request "/" should match snapshot > /@statusText 1`] = `"OK"`;

exports[`Should match snapshot > Request "/about" should match snapshot > /about@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title >About - Web Widget</title><style >web-widget{display:contents}</style><script type="module" src="/entry.client.ts"></script><script id="state:web-router" type="application/json">{"pathname":"/about","params":{}}</script>
      </head>
      <body>
        <header class="_header_d366fc"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li><li><a href="/fetch">Fetching data</a></li><li><a href="/greet/web-widget">Dynamic routes</a></li><li><a href="/action">Server actions</a></li></ul></nav></header><main class="_main_d366fc"><h1>About</h1><p>The page you&#x27;re looking at is purely static HTML, with no client-side interactivity needed. Because of that, we don&#x27;t need to load any JavaScript. Try viewing the page&#x27;s source, or opening the devtools network panel and reloading.</p></main><footer class="_footer_d366fc"><p>This is a footer</p></footer>
      </body>
    </html>"
`;

exports[`Should match snapshot > Request "/about" should match snapshot > /about@headers 1`] = `
{
  "content-type": "text/html; charset=utf-8",
  "x-module-source": "source://routes/about@route.tsx",
  "x-powered-by": "@web-widget/web-router",
}
`;

exports[`Should match snapshot > Request "/about" should match snapshot > /about@status 1`] = `200`;

exports[`Should match snapshot > Request "/about" should match snapshot > /about@statusText 1`] = `"OK"`;

exports[`Should match snapshot > Request "/action" should match snapshot > /action@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><style >web-widget{display:contents}</style><script type="module" src="/entry.client.ts"></script><script id="state:web-router" type="application/json">{"pathname":"/action","params":{}}</script>
      </head>
      <body>
        <header class="_header_d366fc"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li><li><a href="/fetch">Fetching data</a></li><li><a href="/greet/web-widget">Dynamic routes</a></li><li><a href="/action">Server actions</a></li></ul></nav></header><main class="_main_d366fc"><h1>Server action</h1><!--$--><web-widget import="/routes/action/Echo@widget.tsx" name="Echo" loading="lazy" rendertarget="light" recovering=""><input type="text" value=""/><button>echo</button></web-widget><!--/$--></main><footer class="_footer_d366fc"><p>This is a footer</p></footer>
      </body>
    </html>"
`;

exports[`Should match snapshot > Request "/action" should match snapshot > /action@headers 1`] = `
{
  "content-type": "text/html; charset=utf-8",
  "x-module-source": "source://routes/action/index@route.tsx",
  "x-powered-by": "@web-widget/web-router",
}
`;

exports[`Should match snapshot > Request "/action" should match snapshot > /action@status 1`] = `200`;

exports[`Should match snapshot > Request "/action" should match snapshot > /action@statusText 1`] = `"OK"`;

exports[`Should match snapshot > Request "/api/hello" should match snapshot > /api/hello@body 1`] = `"[{"title":"👋🌍","url":"#id0"},{"title":"你好世界","url":"#id1"},{"title":"Hello world","url":"#id2"},{"title":"こんにちは世界","url":"#id3"},{"title":"مرحبا بالعالم","url":"#id4"},{"title":"헬로월드","url":"#id5"}]"`;

exports[`Should match snapshot > Request "/api/hello" should match snapshot > /api/hello@headers 1`] = `
{
  "content-type": "application/json; charset=utf-8",
  "x-powered-by": "@web-widget/web-router",
}
`;

exports[`Should match snapshot > Request "/api/hello" should match snapshot > /api/hello@status 1`] = `200`;

exports[`Should match snapshot > Request "/api/hello" should match snapshot > /api/hello@statusText 1`] = `""`;

exports[`Should match snapshot > Request "/fetch" should match snapshot > /fetch@body 1`] = `
"<!doctype html>
    <html lang="en">
      <head>
        <script id="shim:es-module">
  if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) {
    self.importShim = Object.assign(
      function importShimProxy() {
        const esModuleShimUrl =
          'https://ga.jspm.io/npm:es-module-shims@1.7.3/dist/es-module-shims.js';
        const promise = new Promise((resolve, reject) => {
          document.head.appendChild(
            Object.assign(document.createElement('script'), {
              src: esModuleShimUrl,
              crossorigin: 'anonymous',
              async: true,
              onload() {
                if (!importShim.$proxy) {
                  resolve(importShim);
                } else {
                  reject(
                    new Error(
                      'No globalThis.importShim found:' + esModuleShimUrl
                    )
                  );
                }
              },
              onerror(error) {
                reject(error);
              },
            })
          );
        });
        return promise.then((importShim) => importShim(...arguments));
      },
      {
        $proxy: true,
      }
    );
  }
</script> <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title >Fetching data - Web Widget</title><style >web-widget{display:contents}</style><script type="module" src="/entry.client.ts"></script><script id="state:web-router" type="application/json">{"pathname":"/fetch","params":{}}</script>
      </head>
      <body>
        <header class="_header_d366fc"><nav><ul><li><a href="/">Home</a></li><li><a href="/about">About</a></li><li><a href="/fetch">Fetching data</a></li><li><a href="/greet/web-widget">Dynamic routes</a></li><li><a href="/action">Server actions</a></li></ul></nav></header><main class="_main_d366fc"><h1>Fetching data</h1><ul><li>👋🌍</li><li>你好世界</li><li>Hello world</li><li>こんにちは世界</li><li>مرحبا بالعالم</li><li>헬로월드</li></ul><p>Source:<a href="/api/hello" target="_blank">/api/hello</a></p></main><footer class="_footer_d366fc"><p>This is a footer</p></footer>
      </body>
    </html>"
`;

exports[`Should match snapshot > Request "/fetch" should match snapshot > /fetch@headers 1`] = `
{
  "content-type": "text/html; charset=utf-8",
  "x-module-source": "source://routes/fetch@route.tsx",
  "x-powered-by": "@web-widget/web-router",
}
`;

exports[`Should match snapshot > Request "/fetch" should match snapshot > /fetch@status 1`] = `200`;

exports[`Should match snapshot > Request "/fetch" should match snapshot > /fetch@statusText 1`] = `"OK"`;
